Dependencies cannot use `_%%` in constraints

I have a build file with some scala dependencies in there, and want to cross-compile the module. 
My dependencies are using [constraints](https://docs.gradle.org/current/userguide/dependency_constraints.html#sec:adding-constraints-transitive-deps) to set the version, see:

```kotlin
dependencies {
    implementation("org.scala-lang:scala-library")
    testImplementation("org.scalatest:scalatest_%%")

    constraints {
        dependencies {
            implementation("org.scala-lang:scala-library:%scala-version%")
            testImplementation("org.scalatest:scalatest_%%:3.2.3")
        }
    }
```

This results in the following error

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':myproject:compileTestScala'.
> Could not resolve all files for configuration ':myproject:testCompileClasspath'.
   > Could not resolve org.scalatest:scalatest_%%.
     Required by:
         project :myproject
      > Invalid format: 'org.scalatest:scalatest_2.13:'. Group, name and version cannot be empty. Correct example: 'org.gradle:gradle-core:1.0'
```

--- a/src/main/groovy/com/adtran/ScalaMultiVersionPlugin.groovy
+++ b/src/main/groovy/com/adtran/ScalaMultiVersionPlugin.groovy
@@ -15,11 +15,9 @@
  */
 package com.adtran
 
-import org.gradle.api.GradleException
-import org.gradle.api.Plugin
-import org.gradle.api.Project
-import org.gradle.api.XmlProvider
-import org.gradle.api.artifacts.Dependency
+import groovy.transform.stc.ClosureParams
+import groovy.transform.stc.FromString
+import org.gradle.api.*
 import org.gradle.api.artifacts.component.ModuleComponentSelector
 import org.gradle.api.artifacts.component.ProjectComponentSelector
 import org.gradle.api.artifacts.result.ResolvedDependencyResult
@@ -39,7 +37,8 @@ class ScalaMultiVersionPlugin implements Plugin<Project> {
         this.project = project
         setExtension()
         calculateScalaVersions()
-        setResolutionStrategy()
+        updateDependencyConstraints()
+        updateDependencies()
         addTasks()
         addSuffixToJars()
         resolvePomDependencies()
@@ -82,32 +81,68 @@ class ScalaMultiVersionPlugin implements Plugin<Project> {
         project.ext.defaultScalaVersions = parseScalaVersionList("defaultScalaVersions")
     }
 
-    private String replaceScalaVersions(Dependency dependency) {
-        def newName = dependency.name.replace(
+    private String replaceSuffixPlaceholder(String str) {
+        str.replace(
             project.scalaMultiVersion.scalaSuffixPlaceholder,
             project.ext.scalaSuffix
-        ).replace(
-            project.scalaMultiVersion.scalaVersionPlaceholder,
-            project.ext.scalaVersion
         )
-        def newVersion = dependency.version.replace(
-            project.scalaMultiVersion.scalaVersionPlaceholder,
-            project.ext.scalaVersion)
-        return "$dependency.group:$newName:$newVersion"
     }
 
-    private void setResolutionStrategy() {
-        project.configurations.all {
-            withDependencies { dependencies ->
-                dependencies.findAll { Dependency dependency ->
-                    [ project.scalaMultiVersion.scalaSuffixPlaceholder, project.scalaMultiVersion.scalaVersionPlaceholder ].any {
-                        "${dependency.name}:${dependency.version}".contains(it)
-                    }
-                }.forEach { Dependency dependency ->
-                    def newTarget = replaceScalaVersions(dependency)
-                    dependencies.add(project.dependencies.create(newTarget))
-                    dependencies.remove(dependency)
-                }
+    private String replaceVersionPlaceholder(String str) {
+        str.replace(
+            project.scalaMultiVersion.scalaVersionPlaceholder,
+            project.ext.scalaVersion)
+    }
+
+    private String prepareObjectNotion(domainObject) {
+        def newName = replaceVersionPlaceholder(replaceSuffixPlaceholder(domainObject.name))
+        if (domainObject.version == null) {
+            return "$domainObject.group:$newName"
+        }
+        def newVersion = replaceVersionPlaceholder(domainObject.version)
+        return "$domainObject.group:$newName:$newVersion"
+    }
+
+    private <T> void updateDomainObjectSet(DomainObjectSet<T> domainObjectSet, @ClosureParams(value = FromString, options = ["String, T"]) Closure<T> copyObject) {
+        domainObjectSet.findAll {domainObject ->
+            [ project.scalaMultiVersion.scalaSuffixPlaceholder, project.scalaMultiVersion.scalaVersionPlaceholder ].any {
+                "${domainObject.name}:${domainObject.version}".contains(it)
+            }
+        }.forEach {domainObj ->
+            def newNotion = prepareObjectNotion(domainObj)
+            def newObj = copyObject(newNotion, domainObj)
+            domainObjectSet.add(newObj)
+            domainObjectSet.remove(domainObj)
+        }
+    }
+
+    private void updateDependencyConstraints() {
+        project.afterEvaluate {
+            project.configurations.configureEach {
+                updateDomainObjectSet(it.dependencyConstraints, { notion, initialConstraint ->
+                    project.dependencies.constraints.create(notion, {constraint ->
+                        def oldVersionConstraint = initialConstraint.versionConstraint
+                        constraint.version {
+                            it.strictly replaceVersionPlaceholder(oldVersionConstraint.strictVersion)
+                            it.require replaceVersionPlaceholder(oldVersionConstraint.requiredVersion)
+                            it.prefer replaceVersionPlaceholder(oldVersionConstraint.preferredVersion)
+                            def rejectedVersions = oldVersionConstraint.rejectedVersions.collect { replaceVersionPlaceholder(it) } as String[]
+                            it.reject rejectedVersions
+                        }
+                        constraint.because initialConstraint.reason
+                    })
+                })
+            }
+        }
+    }
+
+    private void updateDependencies() {
+        project.configurations.configureEach { conf ->
+            conf.withDependencies { dependencies ->
+                updateDomainObjectSet(dependencies, {
+                    notion, initialDependency ->
+                        project.dependencies.create(notion)
+                })
             }
         }
     }

--- a/src/test/groovy/unit/TestScalaMultiVersionPlugin.groovy
+++ b/src/test/groovy/unit/TestScalaMultiVersionPlugin.groovy
@@ -248,9 +248,17 @@ class TestScalaMultiVersionPlugin extends GroovyTestCase implements SimpleProjec
                 assert !metadataJsonStr.contains("_%%")
                 assert !metadataJsonStr.contains("%scala_version%")
                 def metadataJson = new JsonSlurper().parseText(metadataJsonStr)
-                def runtimeDependencies = metadataJson.variants.find { it.name == "runtimeElements" }.dependencies
+                def runtimeElementsVariant = metadataJson.variants.find { it.name == "runtimeElements" }
+                def runtimeDependencies = runtimeElementsVariant.dependencies
+                assert runtimeDependencies != null
                 assert runtimeDependencies.find { it.module == "scala-library" }.version.requires == ver
                 assert runtimeDependencies.find { it.module == "fake-scala-dep_$base" } != null
+                def runtimeDependencyConstraints = runtimeElementsVariant.dependencyConstraints
+                assert runtimeDependencyConstraints != null
+                def fakeScalaDepConstraint = runtimeDependencyConstraints.find { it.module == "fake-scala-dep_$base" }
+                assert fakeScalaDepConstraint != null
+                assert fakeScalaDepConstraint.reason == "foo"
+                assert fakeScalaDepConstraint.version == [requires: "[1.2, 1.3[", prefers: "1.2.4", rejects: ["1.2.5"]]
             }
         }
     }

--- a/src/test/groovy/unit/TestScalaMultiVersionPlugin.groovy
+++ b/src/test/groovy/unit/TestScalaMultiVersionPlugin.groovy
@@ -67,6 +67,49 @@ class TestScalaMultiVersionPlugin extends GroovyTestCase implements SimpleProjec
         deps.each { assert !it.moduleVersion.contains("%scala-version%") }
     }
 
+    void testConstraintResolution() {
+        def project = createProject("2.12.1")
+        def conf = project.configurations.getByName("implementation")
+        conf.dependencyConstraints.each {assert !it.name.contains("_%%") }
+        conf.dependencyConstraints.each {assert !it.name.contains("%scala-version%") }
+        def resolvedConstraint = conf.dependencyConstraints.find { it.name == "fake-scala-dep_2.12" }
+        assert resolvedConstraint != null
+        def versionConstraint = resolvedConstraint.versionConstraint
+        assert versionConstraint.preferredVersion == "1.2.4"
+        assert versionConstraint.requiredVersion == "[1.2, 1.3["
+        assert versionConstraint.rejectedVersions == ["1.2.5"]
+        assert resolvedConstraint.reason == "foo"
+    }
+
+    void testDependencyWithConstraintResolved() {
+        def project = createProject("2.12.1")
+        def conf = project.configurations.getByName("compileClasspath")
+        def resolvedDependencies = conf.resolvedConfiguration.firstLevelModuleDependencies
+        def resolvedDependency = resolvedDependencies.find {it.module.id.name == "fake-scala-dep_2.12" }
+        assert resolvedDependency != null
+        assert resolvedDependency.moduleVersion == "1.2.3"
+    }
+
+    void testUnresolvableConstraint() {
+        def addUnresolvableConstraint = {
+            afterEvaluate {
+                dependencies {
+                    constraints {
+                        implementation("org.scalamacros:fake-compiler-plugin_%scala-version%") {
+                            version {
+                                reject "1.2.3"
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        def project = createProject("2.12.1", addUnresolvableConstraint)
+        def conf = project.configurations.getByName("compileClasspath").resolvedConfiguration.lenientConfiguration
+        assert !conf.unresolvedModuleDependencies.isEmpty()
+        conf.unresolvedModuleDependencies.each {assert !it.selector.name.contains("%scala-version%") }
+    }
+
     void testBadScalaVersions() {
         [ [null, "Must set 'scalaVersions' property."],
           ["", "Invalid scala version '' in 'scalaVersions' property."],

--- a/src/test/groovy/unit/TestScalaMultiVersionPlugin.groovy
+++ b/src/test/groovy/unit/TestScalaMultiVersionPlugin.groovy
@@ -57,14 +57,14 @@ class TestScalaMultiVersionPlugin extends GroovyTestCase implements SimpleProjec
         }
     }
 
-    void testResolutionStrategy() {
+    void testDependencyResolution() {
         def project = createProject("2.12.1")
         def conf = project.configurations.getByName("compileClasspath").resolvedConfiguration.lenientConfiguration
         assert conf.unresolvedModuleDependencies.size() == 0
         def deps = conf.getAllModuleDependencies()
         assert deps.size() == 4
         deps.each { assert !it.name.contains("_%%") }
-        deps.each { assert !it.moduleVersion.contains("scalaVersion") }
+        deps.each { assert !it.moduleVersion.contains("%scala-version%") }
     }
 
     void testBadScalaVersions() {
@@ -155,7 +155,7 @@ class TestScalaMultiVersionPlugin extends GroovyTestCase implements SimpleProjec
             )
         }
         def conf = project.configurations.getByName("compileClasspath").resolvedConfiguration.lenientConfiguration
-        assert conf.unresolvedModuleDependencies.size() == 3
+        assert conf.unresolvedModuleDependencies.size() == 4 // 3 dependencies + 1 constraint
         assert conf.getAllModuleDependencies().size() == 1
     }
 

--- a/testProjects/simpleProject/build.gradle
+++ b/testProjects/simpleProject/build.gradle
@@ -33,8 +33,19 @@ repositories {
 }
 
 dependencies {
+    constraints {
+        implementation("org.fake:fake-scala-dep_%%") {
+            version {
+                strictly '[1.2, 1.3['
+                prefer '1.2.4'
+                reject '1.2.5'
+            }
+            because 'foo'
+        }
+    }
+
     implementation "org.scala-lang:scala-library:%scala-version%"
-    implementation "org.fake:fake-scala-dep_%%:1.2.3"
+    implementation "org.fake:fake-scala-dep_%%"
     implementation "org.fake:fake-java-dep:1.2.3"
     implementation "org.scalamacros:fake-compiler-plugin_%scala-version%:1.2.3"
 }

